<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Initialize genome-level Trellis layout — gtrellis_layout • gtrellis</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Initialize genome-level Trellis layout — gtrellis_layout"><meta property="og:description" content="Initialize genome-level Trellis layout"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">gtrellis</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.31.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-header">Vignettes</li>
    <li>
      <a href="../articles/gtrellis_full.html">Make Genome-level Trellis Graph</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/jokergoo/gtrellis/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Initialize genome-level Trellis layout</h1>
    
    <div class="hidden name"><code>gtrellis_layout.rd</code></div>
    </div>

    <div class="ref-description">
    <p>Initialize genome-level Trellis layout</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gtrellis_layout</span><span class="op">(</span>data <span class="op">=</span> <span class="cn">NULL</span>, category <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>    species <span class="op">=</span> <span class="cn">NULL</span>, nrow <span class="op">=</span> <span class="cn">NULL</span>, ncol <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>    n_track <span class="op">=</span> <span class="fl">1</span>, track_height <span class="op">=</span> <span class="fl">1</span>, track_ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>    track_axis <span class="op">=</span> <span class="cn">TRUE</span>, track_ylab <span class="op">=</span> <span class="st">""</span>, ylab_rot <span class="op">=</span> <span class="fl">90</span>, title <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>    xlab <span class="op">=</span> <span class="st">"Genomic positions"</span>, xaxis <span class="op">=</span> <span class="cn">TRUE</span>, xaxis_bin <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>    equal_width <span class="op">=</span> <span class="cn">FALSE</span>, compact <span class="op">=</span> <span class="cn">FALSE</span>, border <span class="op">=</span> <span class="cn">TRUE</span>, asist_ticks <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    xpadding <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span>, ypadding <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span>, gap <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">1</span>, <span class="st">"mm"</span><span class="op">)</span>,</span>
<span>    byrow <span class="op">=</span> <span class="cn">TRUE</span>, newpage <span class="op">=</span> <span class="cn">TRUE</span>, add_name_track <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    name_fontsize <span class="op">=</span> <span class="fl">10</span>, name_track_fill <span class="op">=</span> <span class="st">"#EEEEEE"</span>,</span>
<span>    add_ideogram_track <span class="op">=</span> <span class="cn">FALSE</span>, ideogram_track_height <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">2</span>, <span class="st">"mm"</span><span class="op">)</span>,</span>
<span>    axis_label_fontsize <span class="op">=</span> <span class="fl">6</span>, lab_fontsize <span class="op">=</span> <span class="fl">10</span>, title_fontsize <span class="op">=</span> <span class="fl">16</span>,</span>
<span>    legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>, legend_side <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"right"</span>, <span class="st">"bottom"</span><span class="op">)</span>,</span>
<span>    padding <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span>, <span class="st">"mm"</span><span class="op">)</span>, remove_chr_prefix <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>data</dt>
<dd><p>a data frame with at least three columns. The first three columns should be genomic categories (e.g. chromosomes),  start positions and end positions. This data frame is used to extract ranges for each genomic category (minimal and maximal positions are taken as the range in the corresponding category).</p></dd>

  <dt>category</dt>
<dd><p>subset of categories. It is also used for ordering.</p></dd>

  <dt>species</dt>
<dd><p>Abbreviations of species. e.g. hg19 for human, mm10 for mouse. If this value is specified, the function will download <code>chromInfo.txt.gz</code> from UCSC ftp automatically. Short scaffolds will be removed if they have obvious different length as others.  Non-normal chromosomes will also be detected and removed. Sometimes this detection is not always correct and if you find chromosomes shown on the plot is not what you expect, set <code>category</code> manually. The argument is passed to <code><a href="https://rdrr.io/pkg/circlize/man/read.chromInfo.html" class="external-link">read.chromInfo</a></code>.</p></dd>

  <dt>nrow</dt>
<dd><p>Number of rows in the layout.</p></dd>

  <dt>ncol</dt>
<dd><p>Number of columns in the layout.</p></dd>

  <dt>n_track</dt>
<dd><p>Number of tracks in each genomic category.</p></dd>

  <dt>track_height</dt>
<dd><p>height of tracks. It should be numeric which means the value is relative and will be scaled into percent, or a <code><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></code> object.</p></dd>

  <dt>track_ylim</dt>
<dd><p>ranges on y axes of tracks. The value can be a vector of length two which means all tracks share same y ranges, or a matrix with two columns, or a vector of length <code>2*n_track</code> which will be coerced into the two-column matrix by rows.</p></dd>

  <dt>track_axis</dt>
<dd><p>whether show y axes for tracks. The value is logical that can be either length one or number of tracks.</p></dd>

  <dt>track_ylab</dt>
<dd><p>labels for tracks on y axes. The value can be either length one or number of tracks.</p></dd>

  <dt>ylab_rot</dt>
<dd><p>value can only be 0 or 90.</p></dd>

  <dt>title</dt>
<dd><p>title of the plot.</p></dd>

  <dt>xlab</dt>
<dd><p>labels on x axes.</p></dd>

  <dt>xaxis</dt>
<dd><p>whether show x axes.</p></dd>

  <dt>xaxis_bin</dt>
<dd><p>bin size for x axes.</p></dd>

  <dt>equal_width</dt>
<dd><p>whether all columns in the layout have the same width. If <code>TRUE</code>, short categories will be extended according to the longest category.</p></dd>

  <dt>compact</dt>
<dd><p>For the catgories which are put in a same row, will they be put compactly without being aligned by columns.</p></dd>

  <dt>border</dt>
<dd><p>whether show borders.</p></dd>

  <dt>asist_ticks</dt>
<dd><p>if axes ticks are added on one side in rows or columns, whether add ticks on the other sides.</p></dd>

  <dt>xpadding</dt>
<dd><p>padding on x axes in each cell. Numeric value means relative ratio corresponding to the cell width.  Use <code><a href="https://rdrr.io/r/base/AsIs.html" class="external-link">I</a></code> to set it as absolute value which is measured in the data viewport (the coordinate system corresponding to the real data). Currently you cannot set it as a <code><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></code> object.</p></dd>

  <dt>ypadding</dt>
<dd><p>padding on y axes in each cell. Only numeric value is allowed currently.</p></dd>

  <dt>gap</dt>
<dd><p>0 or a <code><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></code> object. If it is length two, the first element corresponds to the gaps between rows and the second corresponds to the gaps between columns.</p></dd>

  <dt>byrow</dt>
<dd><p>arrange categories (e.g. chromosomes) by rows or by columns in the layout.</p></dd>

  <dt>newpage</dt>
<dd><p>whether call <code><a href="https://rdrr.io/r/grid/grid.newpage.html" class="external-link">grid.newpage</a></code> to create a new page.</p></dd>

  <dt>add_name_track</dt>
<dd><p>whether add a pre-defined name track (insert before the first track). The name track is simply a track which only contains text. The default style of the name track is simple, but users can self define their own by <code><a href="add_track.html">add_track</a></code>.</p></dd>

  <dt>name_fontsize</dt>
<dd><p>font size for text in the name track. Note the font size also affects the height of name track.</p></dd>

  <dt>name_track_fill</dt>
<dd><p>filled color for name track.</p></dd>

  <dt>add_ideogram_track</dt>
<dd><p>whether to add a pre-defined ideogram track (insert after the last track). If the cytoband data for specified species is not available, this argument is ignored. The ideogram track simply contains rectangles with different colors, implemented by <code><a href="add_track.html">add_track</a></code>.</p></dd>

  <dt>ideogram_track_height</dt>
<dd><p>Height of ideogram track. The value should be a <code><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></code> object.</p></dd>

  <dt>axis_label_fontsize</dt>
<dd><p>font size for axis labels.</p></dd>

  <dt>lab_fontsize</dt>
<dd><p>font size for x-labels and y-labels.</p></dd>

  <dt>title_fontsize</dt>
<dd><p>font size for title.</p></dd>

  <dt>legend</dt>
<dd><p>a <code><a href="https://rdrr.io/r/grid/grid-defunct.html" class="external-link">grob</a></code> or <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Legends-class.html" class="external-link">Legends-class</a></code> object, or a list of them.</p></dd>

  <dt>legend_side</dt>
<dd><p>side of the legend</p></dd>

  <dt>padding</dt>
<dd><p>padding of the plot. Elements correspond to bottom, left, top, right paddings.</p></dd>

  <dt>remove_chr_prefix</dt>
<dd><p>if chromosome names start with 'chr', whether to remove it.</p></dd>


</dl></div>
    <div id="details">
    <h2>Details</h2>
    <p>Genome-level Trellis graph visualizes genomic data conditioned by genomic categories (e.g. chromosomes).
For each genomic category, multiple dimensional data which are represented as tracks describe different features from different
aspects. The <code>gtrellis_layout</code> function arranges genomic categories on the plot in a quite flexible way. Then users
apply <code><a href="add_track.html">add_track</a></code> to add self-defined graphics to the plot track by track.</p>
<p>For more detailed demonstration of the function, please refer to the vignette.</p>
    </div>
    <div id="value">
    <h2>Value</h2>
    

<p>No value is returned.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="add_track.html">add_track</a></code>, <code><a href="add_ideogram_track.html">add_ideogram_track</a></code></p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Zuguang Gu &lt;z.gu@dkfz.de&gt;</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># There is no example</span></span></span>
<span class="r-in"><span><span class="cn">NULL</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NULL</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Zuguang Gu.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  <style>nav[data-toggle='toc'] .nav .nav {display: block;}</style></body></html>

